src/persist-txn/src/txn_cache.rs

```diff
@@ -590,86 +565,85 @@ impl<T: Timestamp + Lattice + TotalOrder + StepForward + Codec64> TxnsCacheState
         } else {
             unreachable!("only +1/-1 diffs are used");
         }
+        self.compact_data_times(&data_id);
+        debug_assert_eq!(self.validate(), Ok(()));
     }
 
-    fn compact_data_times(&mut self) {
-        debug!(
-            "cache compact since={:?} min_writes={:?}",
-            self.since_ts, self.datas_min_write_ts
-        );
-        loop {
-            // Repeatedly grab the data shard with the minimum write_ts while it
-            // (the minimum write_ts) is less_than the since_ts. If that results
-            // in no registration and no writes, forget about the data shard
-            // entirely, so it doesn't sit around in the map forever. This will
-            // eventually finish because each data shard we compact will not
-            // meet the compaction criteria if we see it again.
-            //
-            // NB: This intentionally doesn't compact the registration
-            // timestamps if none of the writes need to be compacted, so that
-            // compaction isn't `O(compact calls * data shards)`.
-            let data_id = match self.datas_min_write_ts.peek() {
-                Some(Reverse((ts, _))) if ts < &self.since_ts => {
-                    let Reverse((_, data_id)) = self.datas_min_write_ts.pop().expect("just peeked");
-                    data_id
-                }
-                Some(_) | None => break,
-            };
-            let times = self
-                .datas
-                .get_mut(&data_id)
-                .expect("datas_min_write_ts should be an index into datas");
+    /// Informs the cache that all registers and forgets less than ts have been
+    /// applied.
+    pub(crate) fn mark_register_applied(&mut self, ts: &T) {
+        self.unapplied_registers
+            .retain(|(_, register_ts)| ts < register_ts);
+        debug_assert_eq!(self.validate(), Ok(()));
+    }
+
+    /// Compact the internal representation for `data_id` by removing all data
+    /// that is not needed to maintain the following invariants:
+    ///
+    ///   - The latest write and registration for each shard are kept in
+    ///     `self.datas`.
+    ///   - All unapplied writes and registrations are kept in `self.datas`.
+    ///   - All writes in `self.datas` are contained by some registration in
+    ///     `self.datas`.
+    fn compact_data_times(&mut self, data_id: &ShardId) {
```

Comment:
I like the solution of making `data_id` a parameter and only calling this when necessary for that data shard, nice!